sql: fix implicit record types for virtual tables

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/90871

The fixes for the pg_type(oid) index and the ability to cast to an
implicit record type for a virtual table were quite related to each
other, since it had to do with how the ID for the implicit type is
calculated.

Release note (bug fix): Fixed a bug where point loookups on the
pg_catalog.pg_type table would fail to find the implicit record type
that gets created for tables in the pg_catalog, information_schema, and
crdb_internal schemas.

Release note (bug fix): Fixed a bug that prevented the usage of implicit
record types for tables in the pg_catalog, information_schema, and
crdb_internal schemas.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/90924)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> Reviewed 2 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90924)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)

___
*[`pkg/sql/catalog/typedesc/table_implicit_record_type.go` line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90924#-NFeXOSL1Qc3JTnWvjRa:-NFeXOSM5ER668wmtXBr:b-yujcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c93e557a896a0368dfe6e5a705c9ac28188a0580/pkg/sql/catalog/typedesc/table_implicit_record_type.go#L94)):*
> ```Go
> 		tablePrivs = catpb.NewPrivilegeDescriptor(
> 			username.PublicRoleName(), privilege.List{privilege.USAGE}, privilege.List{}, username.NodeUserName(),
> 		)
> ```

1) Is there a way we can allocate this eagerly and avoid the allocatione very time we need to access the privileges? This is one of those thing that I could see showing up in profiles.
2) How does this interact with the virtual table privileges? 



_Code quote:_
```Go
		// Virtual tables don't have privileges attached to them, but we can
		// fake it here so that the implicit record type gets the USAGE privilege.
		tablePrivs = catpb.NewPrivilegeDescriptor(
			username.PublicRoleName(), privilege.List{privilege.USAGE}, privilege.List{}, username.NodeUserName(),
		)
```


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90924)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mgartner)

___
*[`pkg/sql/catalog/typedesc/table_implicit_record_type.go` line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90924#-NFeXOSL1Qc3JTnWvjRa:-NFkaNyN-khmZzoD9Vls:b-m9vy7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c93e557a896a0368dfe6e5a705c9ac28188a0580/pkg/sql/catalog/typedesc/table_implicit_record_type.go#L94)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

1) Is there a way we can allocate this eagerly and avoid the allocatione very time we need to access the privileges? This is one of those thing that I could see showing up in profiles.
2) How does this interact with the virtual table privileges? 
</blockquote></details>

actually, it's looking like the correct fix here is to remove the implicit types for virtual tables from `pg_type` completely. they don't actually exist.

```
root@localhost:26257/defaultdb> select (1, 'cat', NULL, NULL)::pg_namespace;
ERROR: type "pg_namespace" does not exist
SQLSTATE: 42704

root@localhost:26257/defaultdb> select (1, 'cat', NULL, NULL)::pg_catalog.pg_namespace;
ERROR: type "pg_catalog.pg_namespace" does not exist
SQLSTATE: 42704

root@localhost:26257/defaultdb> create table my_namespace(a oid, b text, c text, d text);
CREATE TABLE

root@localhost:26257/defaultdb> select (1, 'cat', NULL, NULL)::my_namespace;
  my_namespace
----------------
  (1,cat,,)
(1 row)
```


<!-- Sent from Reviewable.io -->

rafiss committed:
sql: fix implicit record types for virtual tables

The fixes for the pg_type(oid) index and the ability to cast to an
implicit record type for a virtual table were quite related to each
other, since it had to do with how the ID for the implicit type is
calculated.

Release note (bug fix): Fixed a bug where point loookups on the
pg_catalog.pg_type table would fail to find the implicit record type
that gets created for tables in the pg_catalog, information_schema, and
crdb_internal schemas.

Release note (bug fix): Fixed a bug that prevented the usage of implicit
record types for tables in the pg_catalog, information_schema, and
crdb_internal schemas.

ajwerner reviewed (commented):
You know postgres has these entries, yeah? 

> Reviewed 2 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90924#-:-NFnNi8xDMdAPAPf_xEa:b6iufj1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
It seems like an alternative is to make the virtual tables behave more like regular table

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90924#-:-NFnNv3-54_HI-sbDpcr:bjc9wbj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
ok yeah, i'll make the virtual tables behave more regularly instead.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90924#-:-NFoXHE-5Boba2PPqP3Q:by2r2fu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mgartner)

___
*[`pkg/sql/catalog/typedesc/table_implicit_record_type.go` line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90924#-NFeXOSL1Qc3JTnWvjRa:-NFoXL9R-wwDTsHO1mSN:b-c8dpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c93e557a896a0368dfe6e5a705c9ac28188a0580/pkg/sql/catalog/typedesc/table_implicit_record_type.go#L94)):*
i'm going back to supporting virtual tables here.

> Is there a way we can allocate this eagerly and avoid the allocatione very time we need to access the privileges? This is one of those thing that I could see showing up in profiles.

i can eagerly allocate it, though given that we haven't seen any of the `catpb.New*PrivilegeDescriptor` calls show up in profiles, i don't know if it's a big concern.

> How does this interact with the virtual table privileges?

that's tricky. it doesn't right now, and this package can't even access the synthetic privilege logic. i think it's fine to say that implicit record types for virtual tables are all hardcoded to have USAGE privileges and it can't be modified. that's where we were with vtables themselves up until v22.2, and unlike the vtable, there is no sensitive data in an implicit record for a vtable that seems worth restricting access to.


<!-- Sent from Reviewable.io -->

rafiss commented:
this is ready for another look

ajwerner reviewed (commented):
> Reviewed 4 of 10 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90924)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)

___
*[`pkg/sql/catalog/resolver/resolver.go` line 240 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/90924#-NFtUEem1JEVp3TvOBeE:-NFtUEen9FzNKYV7apSj:b-ffwg16) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada8b32095d3e207a8d8ed11d1370282b4bb4199/pkg/sql/catalog/resolver/resolver.go#L240)):*
> ```Go
> 			return nil, prefix, err
> 		}
> 		return implicitTyp, prefix, nil
> ```

:thinking: this hints to me that we're doing this in the wrong place in the descs.Collection. Do you need this? I'd expect that this would already happen in `descs.Collection.GetObjectByName`



_Code quote:_
```Go
	switch lookupFlags.DesiredObjectKind {
	case tree.TypeObject:
		typ, isType := obj.(catalog.TypeDescriptor)
		if isType {
			return typ, prefix, nil
		}
		table, isTable := obj.(catalog.TableDescriptor)
		if !isTable {
			return nil, prefix, sqlerrors.NewUndefinedTypeError(getResolvedTn())
		}
		implicitTyp, err := typedesc.CreateImplicitRecordTypeFromTableDesc(table)
		if err != nil {
			return nil, prefix, err
		}
		return implicitTyp, prefix, nil
```


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90924)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mgartner)

___
*[`pkg/sql/catalog/resolver/resolver.go` line 240 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/90924#-NFtUEem1JEVp3TvOBeE:-NFtdb9jD4E4NaR9lLPM:b-qp7lyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada8b32095d3e207a8d8ed11d1370282b4bb4199/pkg/sql/catalog/resolver/resolver.go#L240)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

:thinking: this hints to me that we're doing this in the wrong place in the descs.Collection. Do you need this? I'd expect that this would already happen in `descs.Collection.GetObjectByName`
</blockquote></details>

nice catch! i removed it and found this isn't needed. i could have sworn i only added it in after stepping through and seeing a regression test fail here, but maybe i made the `GetObjectByName` after and that fixed it better.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> very cool 

> Reviewed 4 of 10 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90924#-:-NFtfFTp3g2KqET4DotH:bhhjxvg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

rafiss commented:
bors r=ajwerner
